### DISCUSSION TITLE
subscription to user status

### DISCUSSION THREAD

**Discussion by rraammiinn** [pocketbase/pocketbase#3437] [Q&A]
can you add a feature for subscribing to users offline/online status ?

**Comment by ganigeorgiev** [comment]
If you are asking for presence-like functionality, then No because it is not clear who should have access to the presence information, whether there would need to be some grouping (eg. users from a specific "room"), etc. and for now it is left to the developers to handle it on their own based on their specific use case (_see also https://github.com/pocketbase/pocketbase/discussions/3378 and https://github.com/pocketbase/pocketbase/discussions/845_)

**Reply by rraammiinn** [reply]
I'm using it for creating a simple chat app but for checking users offline/online status l should constantly update user's lastvisited field and subscribing to this field for checking user status and I think it's pretty inefficient .

**Reply by ganigeorgiev** [reply]
I'm not entirely sure that I understand what do you mean, but again there are just too many options that will depend on the specific use case and it is difficult to generalize it (eg. who should have access to the offline/online status - everyone, all authenticated users or only the users from a specific chat room?) and it is better left to the developers to handle it on their own for example using one of the suggestions from https://github.com/pocketbase/pocketbase/discussions/3378.

**Reply by rraammiinn** [reply]
why online status couldn't be a field like other fields in users table rows ?

**Reply by rraammiinn** [reply]
or like email field which you can specify if it's public or not ?

**Reply by rraammiinn** [reply]
you did it for the email field so what's the difference ?

**Reply by ganigeorgiev** [reply]
Online status for what exactly? What should be the behavior if someone wants to show the "online/offline" status only for a specific subscription topic/collection?

> or like email field which you can specify if it's public or not ?

But this toggle option is global and it is managed by the user. It can quickly get complicated if you want presence per "group"?

The behavior is just too ambiguous and will depend on the specific use case.

**Reply by rraammiinn** [reply]
no I don't mean presence in group I mean simply a user that is connecting to server and interacting with it like subscribing to something .

**Reply by rraammiinn** [reply]
the problem is I can define a boolean field called isOnline for example and set it to true when user first connects to server but I can't set it back to false when user closes the browser so I did it through defining a field called lastVisited which is a date field but it's very inefficient .

**Reply by ganigeorgiev** [reply]
I understand but even for your simple use case you'll still need first to have access to the users collection to list them and second to subscribe to the entire collection to track the changes.

And again - this works only for your specific use case. The next person will want something else.

I may eventually revisit this in the future since it was requested several times, but for now the recommended approach is to the use the [Realtime API hooks](https://pocketbase.io/docs/go-event-hooks/#realtime-api-hooks) and manage a custom topic on your own (_or toggle the state of a local field_).

**Reply by rraammiinn** [reply]
thanks .
I wish there was something like cloud functions in pocketbase .
that way much of the problems would be solved easily .

**Reply by ganigeorgiev** [reply]
You can listen to the realtime hooks using JS too but keep in mind that it is not a browser/node environment and has some limitations (see [JSVM Caveats and limitations](https://pocketbase.io/docs/js-overview/#caveats-and-limitations)).

Here is an example `pb_hooks/main.pb.js` file that will toggle a "online" field (_I haven't tested it_):

```js
// note: we listen to the subscribe request and not to the "on connect" 
// because the authentication happens during the subscription submission
onRealtimeAfterSubscribeRequest((e) => {
    const record = e.client.get("authRecord")

    if (record && !record.get("online")) {
        record.set("online", true);
        $app.dao().saveRecord(record);
    }
})

onRealtimeDisconnectRequest((e) => {
    const record = e.client.get("authRecord")

    if (record && record.get("online")) {
        record.set("online", false);
        $app.dao().saveRecord(record);
    }
})
```

**Reply by rraammiinn** [reply]
thanks again .
it seems I should study more about pocketbase features .

**Reply by rraammiinn** [reply]
it worked .
now how I can do it in go ?

when I,m trying to do the same thing in go the return value of ( e.Client.Get("authRecord") ) is of type any and when I say ( e.Client.Get("authRecord") ) it says ( type any has no field or method Get) .

**Reply by ganigeorgiev** [reply]
Are you sure that the error is referring to the `e.Client` and not something else?
Note that when using Go, `e.Client.Get()` returns `any` and you'll have to type assert it to the correct type before being able to access the record fields.

In any case, below is a Go translation of the above hooks:

```go
// note: we listen to the subscribe request and not to the "on connect" 
// because the authentication happens during the subscription submission
app.OnRealtimeAfterSubscribeRequest().Add(func(e *core.RealtimeSubscribeEvent) error {
    record, _ := e.Client.Get(apis.ContextAuthRecordKey).(*models.Record)
    if record != nil && !record.GetBool("online") {
        record.Set("online", true)
        return app.Dao().SaveRecord(record)
    }
    return nil
})

app.OnRealtimeDisconnectRequest().Add(func(e *core.RealtimeDisconnectEvent) error {
    record, _ := e.Client.Get(apis.ContextAuthRecordKey).(*models.Record)
    if record != nil && record.GetBool("online") {
        record.Set("online", false)
        return app.Dao().SaveRecord(record)
    }
    return nil
})
```

If the above doesn't help, please elaborate and paste your code that you've tried.

**Reply by rraammiinn** [reply]
thanks .
I,m using js hooks for development but for production I think go is beter .
